Add linear regression

SarahAlidoost opened the pull request:
closes #2 
closes #19

fnattino commented on src/fowt_ml/linear_models.py:
```diff
@@ -0,0 +1,100 @@
+@dataclass
+class Estimator:
+    name: str
+    func: callable
```
If I understand correctly, this is actually assigned an estimator instance (something that implements `estimator.fit()` and `estimator.predict()`) - so not a callable function, right?

SarahAlidoost replied:
I refactored this. Now it uses `BaseEstimator`

fnattino commented on src/fowt_ml/linear_models.py:
```diff
@@ -0,0 +1,100 @@
+        "mean_absolute_error": sm.mean_absolute_error,
+    }
+
+    def __init__(self, estimator: str | Estimator, kwargs: dict = None) -> None:
```
Maybe instead of `kwargs` either use `**kwargs` (i.e. with unpacking) or use a more descriptive name (maybe `params`?)?

Also, I am not 100% sure, but if a parameter has `None` as default, this should be made explicit in the type hint:
```suggestion
    def __init__(self, estimator: str | Estimator, kwargs: dict | None = None) -> None:
```

SarahAlidoost replied:
it is fixed now.

fnattino commented on src/fowt_ml/linear_models.py:
```diff
@@ -0,0 +1,100 @@
+            raise ValueError("model must be a string or a Estimator instance.")
+
+        # fill the estimator with the kwargs
+        if kwargs:
+            for key, value in kwargs.items():
+                setattr(self.estimator.func, key, value)
```
Estimator objects in scikit-learn should have a `set_params` method (see [link](https://scikit-learn.org/dev/developers/develop.html#estimators)), so maybe one could do:
```suggestion
        self.estimator.func.set_params(**kwargs)
```

SarahAlidoost replied:
cool, I didn't know the method. thanks for the suggestion, it is fixed now.

fnattino reviewed (commented):
Looks good @SarahAlidoost, just a few minor comment. 

One think that I have thought of right now while looking at the code is data normalization, which I think is not implemented but maybe should be included as part of the data preparation (maybe in the pipeline)?

SarahAlidoost commented:
> Looks good @SarahAlidoost, just a few minor comment.
> 
> One think that I have thought of right now while looking at the code is data normalization, which I think is not implemented but maybe should be included as part of the data preparation (maybe in the pipeline)?

Thanks for the review and suggestions. I addressed your comments, mainly the implementation is refactored in line with random forest implementation. Please let me know if we can merge this PR.

fnattino commented on src/fowt_ml/linear_models.py:
```diff
@@ -0,0 +1,65 @@
+        Returns:
+            float: the scoring value
+        """
+        self.model = self.estimator.fit(x_train, y_train)
```
As I also commented in #16, I think `estimator.fit()` updates the estimator in place and then self-return - so I think `self.model` and `self.estimator` here would be identical..

fnattino reviewed (approved):
Looks good @SarahAlidoost  - I have left only one comment, feel free to address it or not  and merge it afterwards :)
